### PR TITLE
Don't convert entry.Name twice when accessing updateIndex

### LINKED ISSUE #82: Can not delete directory from zip archive bug
ICSharpCode.SharpZipLib version - 0.86.0.518.

Code to demonstrate the bug:

``` csharp
namespace SharpZipLibDeleteFolderBug
{
    using ICSharpCode.SharpZipLib.Zip;

    using NUnit.Framework;

    [TestFixture]
    public class SharpZipLibDeleteFolderBugDemo
    {
        [Test]
        public void Bug()
        {
            var zipFile = ZipFile.Create("SomeFile.zip");
            CreateDirectory(zipFile, "SomeDirectory");

            DeleteDirectory(zipFile, "SomeDirectory");
        }

        private void DeleteDirectory(ZipFile zipFile, string directoryName)
        {
            zipFile.BeginUpdate();

            var zipEntry = zipFile.GetEntry(directoryName + "/");
            Assert.That(zipEntry, Is.Not.Null); // passed - directory was successfully create

            zipFile.Delete(zipEntry); // !!! throws ICSharpCode.SharpZipLib.Zip.ZipException : Cannot find entry to delete

            zipFile.CommitUpdate();            
        }

        private static void CreateDirectory(ZipFile zipFile, string directoryName)
        {
            zipFile.BeginUpdate();
            zipFile.AddDirectory(directoryName);
            zipFile.CommitUpdate();
        }
    }
}
```

The buggy method is ZipFile.FindExistingUpdate:

``` csharp
        int FindExistingUpdate(ZipEntry entry)
        {
            int result = -1;
            string convertedName = GetTransformedFileName(entry.Name); // !!! BUG is here

            if (updateIndex_.ContainsKey(convertedName)) {
                result = (int)updateIndex_[convertedName];
            }

                        ...
            return result;
        }
```

Method FindExistingUpdate ignores the fact that entry can be not only file entry but also a directory entry. Right implementation should be:

``` csharp
        int FindExistingUpdate(ZipEntry entry)
        {
            int result = -1;
            string convertedName;
            if (entry.IsDirectory)
                convertedName = GetTransformedDirectoryName(entry.Name);
            else 
                convertedName = GetTransformedFileName(entry.Name);

            if (updateIndex_.ContainsKey(convertedName)) {
                result = (int)updateIndex_[convertedName];
            }

                        ...
            return result;
        }
```

Method ZipFile.Delete(string fileName) also has the same bug.

### PR DESCRIPTION
**Pull request by AqlaSolutions**
I found out that deleting empty directory entry from ZipFile and then adding it back causes an exception while trying to add to `updateIndex` dictionary - because checking for existing element uses transformed `entity.Name` while adding uses its original value. Also directory name gets transformed as file name and ending slash gets removed therefore those too values are not same especially in directory case.
`Entry.Name` is already transformed file name so it shouldn't be transformed twice. So I replaced all such accesses to use entry.Name or transformed string-filename.

Fixes #82

_I certify that I own, and have sufficient rights to contribute, all source code and related material intended to be compiled or integrated with the source code for the SharpZipLib open source product (the "Contribution"). My Contribution is licensed under the MIT License._

### CODE REVIEW

**Comment by Numpsy** [comment]
is this covering similar ground to #309? (haven't looked at the issue, just a query as they look to be changing similar things)

**Comment by piksel** [comment]
Yes, that PR was supposed to fix that issue, but it's been stale for almost a year.

File: src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs [hunk 1]
```diff
@@ -1620,7 +1620,7 @@ private void AddUpdate(ZipUpdate update)
 		{
 			contentsEdited_ = true;
 
-			int index = FindExistingUpdate(update.Entry.Name);
+			int index = FindExistingUpdate(update.Entry.Name, true);
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]: I try to add labels for all booleans to make the code more descriptive (without relying on an IDE).
```suggestion
			int index = FindExistingUpdate(update.Entry.Name, isEntityName: true);
```

File: src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs [hunk 2]
```diff
@@ -2553,11 +2549,11 @@ private int FindExistingUpdate(ZipEntry entry)
 			return result;
 		}
 
-		private int FindExistingUpdate(string fileName)
+		private int FindExistingUpdate(string fileName, bool isEntityName = false)
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]: Is this supposed to be `isEntryName`? Not sure why the word "entity" is used here?

Comment (AqlaSolutions) [reply src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]: You are right, it should be `isEntryName`

Comment (piksel) [reply src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]:
```suggestion
		private int FindExistingUpdate(string fileName, bool isEntryName = false)
		{
			int result = -1;

			string convertedName = GetTransformedFileName(fileName);
			string convertedName = !isEntryName ? GetTransformedFileName(fileName) : fileName;
```

File: src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs [hunk 3]
```diff
@@ -2553,11 +2549,11 @@ private int FindExistingUpdate(ZipEntry entry)
...
 		{
 			int result = -1;
 
-			string convertedName = GetTransformedFileName(fileName);
+			string convertedName = !isEntityName ? GetTransformedFileName(fileName) : fileName;
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]:
```suggestion
			string convertedName = !isEntryName ? GetTransformedFileName(fileName) : fileName;
```

File: src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs [hunk 4]
```diff
@@ -1620,7 +1620,7 @@ private void AddUpdate(ZipUpdate update)
 		{
 			contentsEdited_ = true;
 
-			int index = FindExistingUpdate(update.Entry.Name);
+			int index = FindExistingUpdate(update.Entry.Name, isEntityName: true);
```
Comment (piksel) [inline src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs]:
```suggestion
			int index = FindExistingUpdate(update.Entry.Name, isEntryName: true);
```

**Pull request merged**